sound/soc/sof/trace.c

```diff
@@ -132,16 +132,26 @@ static int trace_debugfs_create(struct snd_sof_dev *sdev)
 
 int snd_sof_init_trace_ipc(struct snd_sof_dev *sdev)
 {
-	struct sof_ipc_dma_trace_params params;
+	struct sof_ipc_fw_ready *ready = &sdev->fw_ready;
+	struct sof_ipc_fw_version *v = &ready->version;
+	struct sof_ipc_dma_trace_params_ext params;
 	struct sof_ipc_reply ipc_reply;
 	int ret;
 
 	if (sdev->dtrace_is_enabled || !sdev->dma_trace_pages)
 		return -EINVAL;
 
 	/* set IPC parameters */
-	params.hdr.size = sizeof(params);
-	params.hdr.cmd = SOF_IPC_GLB_TRACE_MSG | SOF_IPC_TRACE_DMA_PARAMS;
+	params.hdr.cmd = SOF_IPC_GLB_TRACE_MSG;
+	/* PARAMS_EXT is only supported from ABI 3.7.0 onwards */
+	if (v->abi_version >= SOF_ABI_VER(3, 7, 0)) {
+		params.hdr.size = sizeof(struct sof_ipc_dma_trace_params_ext);
+		params.hdr.cmd |= SOF_IPC_TRACE_DMA_PARAMS_EXT;
+		params.timestamp = ktime_to_us(ktime_get());
```

Comment:
would be good to comment what time based is used here?
Also wondering if when the trace lasts for a very long time how the firmware will remain in alignment with e.g. kernel logs. Is there a need for an update or realignment from time to time?

Reply:
@plbossart I will add comments here and at trace.h. I am not sure if the clock on kernel and firmware are skew. If not, I think we will only need to align it when the timer on firmware is reset.

Reply:
@bardliao The change looks very useful and I had the same comment on the timestamp. The timebase is not documented in the uapi headers, but the unit would seem to be be DSP ticks.

So shouldn't you convert kernel time into DSP ticks (19.2 or 24 XTAL, sof-logger -c arg) when filling the params.timestamp to ensure the clocks share the same timebase?

@plbossart 
Skewing will happen as well (most machines will sync via NTP/cellular to an external clock), but that is perhaps a lesser concern and could be fixed with period timebase updates. In practical scenarios, I doubt this is an issue as runtime PM will happen frequently.

Reply:
> @bardliao The change looks very useful and I had the same comment on the timestamp. The timebase is not documented in the uapi headers, but the unit would seem to be be DSP ticks.
> 
> So shouldn't you convert kernel time into DSP ticks (19.2 or 24 XTAL, sof-logger -c arg) when filling the params.timestamp to ensure the clocks share the same timebase?

Good point. The current unit is ns and the expectation was that the firmware would convert to its internal base which is currently not known to the kernel. we could alternatively pass the information directly in a unit that's native to the DSP, but we would first need to know the frequency. I am not sure if the firmware reports anywhere what its internal tick is?

> 
> @plbossart
> Skewing will happen as well (most machines will sync via NTP/cellular to an external clock), but that is perhaps a lesser concern and could be fixed with period timebase updates. In practical scenarios, I doubt this is an issue as runtime PM will happen frequently.

Reply:
@kv2019i Agree with @plbossart it will be easier to convert the time base in firmware side. Kernel don't need to know the DSP ticks base.

Reply:
@bardliao  Ack, but that indeed requires changes to FW side as the previous implementation already got merged. But I see you have PR1358 already running, so that seems ok.